File: q0402/q0402.java
class Solution {
    public String removeKdigits(String num, int k) {
        Deque<Character> stack = new ArrayDeque<>();
        for (int i = 0; i < num.length(); i++) {
            char digit = num.charAt(i);
            while (k > 0 && !stack.isEmpty() && digit < stack.peek()) {
                stack.pop();
                k--;
            }
            stack.push(digit);
        }

        while (k > 0) {
            stack.pop();
            k--;
        }

        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) {
            sb.append(stack.pop());
        }
        sb.reverse();
        while (sb.length() > 1 && sb.charAt(0) == '0') {
            sb.deleteCharAt(0);
        }

        return sb.length() == 0 ? "0" : sb.toString();
    }
}

File: q0402/q0402.cpp
string removeKdigits(string num, int k)
{
    vector<char> stack;
    if (k == num.size())
        return "0";
    int i = 0;
    while (i < num.size())
    {
        int n = stack.size();
        while (k && n && stack[n - 1] > num.at(i))
        {
            k--;
            stack.pop_back();
            n = stack.size();
        }
        stack.push_back(num.at(i));
        i++;
    }
    while (k > 0)
    {
        stack.pop_back();
        k--;
    }

    int n = stack.size();
    string ans = "";
    while (!stack.empty())
    {
        ans = stack[n - 1] + ans;
        n--;
        stack.pop_back();
    }
    while (ans.size() > 1 && ans[0] == '0')
    {
        ans = ans.substr(1);
    }
    return ans;
}